

<!DOCTYPE html>
<html lang="zh-CN" data-default-color-scheme=auto>



<head>
  <meta charset="UTF-8">

  <link rel="apple-touch-icon" sizes="76x76" href="/img/fluid.png">
  <link rel="icon" href="/img/icons/Target.png">
  

  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=5.0, shrink-to-fit=no">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  
  <meta name="theme-color" content="#2f4154">
  <meta name="author" content="ph-bear">
  <meta name="keywords" content="">
  
    <meta name="description" content="intro  太好了！女孩只有在ddl迫近的时候才会想起来写这些、 其实有很多篇但不知为何有种一次就要搞篇大的但其实真的发了的只有一篇捡来的 所以又在捡！  TL;DR 根据需要的fasta内容选软件，本文关心：  上下游n bp 多态性位点所在元件整条序列  不行的话找师兄师姐老师要序列吧、 这是在干什么 我们现在有一个存储了群体变异信息的vcf文件，但由于各种原因和下游应用，需要取出变异位点两">
<meta property="og:type" content="article">
<meta property="og:title" content="vcf2fasta">
<meta property="og:url" content="https://patchaven.github.io/2025/01/15/vcf2fasta/index.html">
<meta property="og:site_name" content="patchaven">
<meta property="og:description" content="intro  太好了！女孩只有在ddl迫近的时候才会想起来写这些、 其实有很多篇但不知为何有种一次就要搞篇大的但其实真的发了的只有一篇捡来的 所以又在捡！  TL;DR 根据需要的fasta内容选软件，本文关心：  上下游n bp 多态性位点所在元件整条序列  不行的话找师兄师姐老师要序列吧、 这是在干什么 我们现在有一个存储了群体变异信息的vcf文件，但由于各种原因和下游应用，需要取出变异位点两">
<meta property="og:locale" content="zh_CN">
<meta property="article:published_time" content="2025-01-14T16:00:00.000Z">
<meta property="article:modified_time" content="2025-02-11T02:44:53.244Z">
<meta property="article:author" content="ph-bear">
<meta name="twitter:card" content="summary_large_image">
  
  
    <meta name="referrer" content="no-referrer-when-downgrade">
  
  
  <title>vcf2fasta - patchaven</title>

  <link  rel="stylesheet" href="https://lib.baomitu.com/twitter-bootstrap/4.6.1/css/bootstrap.min.css" />



  <link  rel="stylesheet" href="https://lib.baomitu.com/github-markdown-css/4.0.0/github-markdown.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/hint.css/2.7.0/hint.min.css" />

  <link  rel="stylesheet" href="https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.css" />



<!-- 主题依赖的图标库，不要自行修改 -->
<!-- Do not modify the link that theme dependent icons -->

<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1749284_5i9bdhy70f8.css">



<link rel="stylesheet" href="//at.alicdn.com/t/c/font_1736178_k526ubmyhba.css">


<link  rel="stylesheet" href="/css/main.css" />


  <link id="highlight-css" rel="stylesheet" href="/css/highlight.css" />
  
    <link id="highlight-css-dark" rel="stylesheet" href="/css/highlight-dark.css" />
  




  <script id="fluid-configs">
    var Fluid = window.Fluid || {};
    Fluid.ctx = Object.assign({}, Fluid.ctx)
    var CONFIG = {"hostname":"patchaven.github.io","root":"/","version":"1.9.8","typing":{"enable":false,"typeSpeed":70,"cursorChar":"_","loop":false,"scope":[]},"anchorjs":{"enable":true,"element":"h1,h2,h3,h4,h5,h6","placement":"left","visible":"hover","icon":""},"progressbar":{"enable":true,"height_px":3,"color":"#29d","options":{"showSpinner":false,"trickleSpeed":100}},"code_language":{"enable":true,"default":"TEXT"},"copy_btn":true,"image_caption":{"enable":true},"image_zoom":{"enable":true,"img_url_replace":["",""]},"toc":{"enable":true,"placement":"right","headingSelector":"h1,h2,h3,h4,h5,h6","collapseDepth":0},"lazyload":{"enable":true,"loading_img":"/img/loading.gif","onlypost":false,"offset_factor":2},"web_analytics":{"enable":true,"follow_dnt":true,"baidu":"63fd1025b507ed16c095540ece0646cf","google":{"measurement_id":null},"tencent":{"sid":null,"cid":null},"leancloud":{"app_id":null,"app_key":null,"server_url":null,"path":"window.location.pathname","ignore_local":false},"umami":{"src":null,"website_id":null,"domains":null,"start_time":"2024-01-01T00:00:00.000Z","token":null,"api_server":null}},"search_path":"/local-search.xml","include_content_in_search":true};

    if (CONFIG.web_analytics.follow_dnt) {
      var dntVal = navigator.doNotTrack || window.doNotTrack || navigator.msDoNotTrack;
      Fluid.ctx.dnt = dntVal && (dntVal.startsWith('1') || dntVal.startsWith('yes') || dntVal.startsWith('on'));
    }
  </script>
  <script  src="/js/utils.js" ></script>
  <script  src="/js/color-schema.js" ></script>
  

  
    <!-- Baidu Analytics -->
    <script async>
      if (!Fluid.ctx.dnt) {
        var _hmt = _hmt || [];
        (function() {
          var hm = document.createElement("script");
          hm.src = "https://hm.baidu.com/hm.js?63fd1025b507ed16c095540ece0646cf";
          var s = document.getElementsByTagName("script")[0];
          s.parentNode.insertBefore(hm, s);
        })();
      }
    </script>
  

  

  

  

  



  
<meta name="generator" content="Hexo 7.3.0"></head>


<body>
  

  <header>
    

<div class="header-inner" style="height: 35vh;">
  <nav id="navbar" class="navbar fixed-top  navbar-expand-lg navbar-dark scrolling-navbar">
  <div class="container">
    <a class="navbar-brand" href="/">
      <strong>patchaven</strong>
    </a>

    <button id="navbar-toggler-btn" class="navbar-toggler" type="button" data-toggle="collapse"
            data-target="#navbarSupportedContent"
            aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <div class="animated-icon"><span></span><span></span><span></span></div>
    </button>

    <!-- Collapsible content -->
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
      <ul class="navbar-nav ml-auto text-center">
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/" target="_self">
                <i class="iconfont icon-home-fill"></i>
                <span>首页</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/archives/" target="_self">
                <i class="iconfont icon-archive-fill"></i>
                <span>归档</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/categories/" target="_self">
                <i class="iconfont icon-category-fill"></i>
                <span>分类</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/about/" target="_self">
                <i class="iconfont icon-user-fill"></i>
                <span>关于</span>
              </a>
            </li>
          
        
          
          
          
          
            <li class="nav-item">
              <a class="nav-link" href="/links/" target="_self">
                <i class="iconfont icon-link-fill"></i>
                <span>链接</span>
              </a>
            </li>
          
        
        
          <li class="nav-item" id="search-btn">
            <a class="nav-link" target="_self" href="javascript:;" data-toggle="modal" data-target="#modalSearch" aria-label="Search">
              <i class="iconfont icon-search"></i>
            </a>
          </li>
          
        
        
          <li class="nav-item" id="color-toggle-btn">
            <a class="nav-link" target="_self" href="javascript:;" aria-label="Color Toggle">
              <i class="iconfont icon-dark" id="color-toggle-icon"></i>
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</nav>

  

<div id="banner" class="banner" parallax=true
     style="background: url('/img/bg/show.jpg') no-repeat center center; background-size: cover;">
  <div class="full-bg-img">
    <div class="mask flex-center" style="background-color: rgba(0, 0, 0, 0.6)">
      <div class="banner-text text-center fade-in-up">
        <div class="h2">
          
            <span id="subtitle">vcf2fasta</span>
          
        </div>

        
          
  <div class="mt-3">
    
    
      <span class="post-meta">
        <i class="iconfont icon-date-fill" aria-hidden="true"></i>
        <time datetime="2025-01-15 00:00" pubdate>
          2025年1月15日 凌晨
        </time>
      </span>
    
  </div>

  <div class="mt-1">
    
      <span class="post-meta mr-2">
        <i class="iconfont icon-chart"></i>
        
          2k 字
        
      </span>
    

    
      <span class="post-meta mr-2">
        <i class="iconfont icon-clock-fill"></i>
        
        
        
          9 分钟
        
      </span>
    

    
    
      
        <span id="busuanzi_container_page_pv" style="display: none">
          <i class="iconfont icon-eye" aria-hidden="true"></i>
          <span id="busuanzi_value_page_pv"></span> 次
        </span>
        

      
    
  </div>


        
      </div>

      
    </div>
  </div>
</div>

</div>

  </header>

  <main>
    
      

<div class="container-fluid nopadding-x">
  <div class="row nomargin-x">
    <div class="side-col d-none d-lg-block col-lg-2">
      

    </div>

    <div class="col-lg-8 nopadding-x-md">
      <div class="container nopadding-x-md" id="board-ctn">
        <div id="board">
          <article class="post-content mx-auto">
            <h1 id="seo-header">vcf2fasta</h1>
            
            
              <div class="markdown-body">
                
                <h2 id="intro">intro</h2>
<blockquote>
<p>太好了！女孩只有在ddl迫近的时候才会想起来写这些、</p>
<p>其实有很多篇但不知为何有种一次就要搞篇大的但其实真的发了的只有一篇捡来的</p>
<p>所以又在捡！</p>
</blockquote>
<h2 id="tldr">TL;DR</h2>
<p>根据需要的fasta内容选软件，本文关心：</p>
<ol type="1">
<li>上下游n bp</li>
<li>多态性位点所在元件整条序列</li>
</ol>
<p>不行的话找师兄师姐老师要序列吧、</p>
<h2 id="这是在干什么">这是在干什么</h2>
<p><strong>我们现在有一个存储了群体变异信息的vcf文件，但由于各种原因和下游应用，需要取出变异位点两侧的序列。</strong></p>
<p>作为存储变异信息的文件格式，vcf文件保存的信息可以说是位点信息。从位点信息到序列信息，需要做到的就是在检测变异时所用到的参考基因组里定位到相应位点，再按照后续对这个序列的操作目的来取包含这个位点的上下游序列(意思是你是想取这个多态性位点所在的基因/CDS或者什么特定的片段，还是你只是关心这个位点的邻居们都是些什么货色)。很简单的思路！不太确定自己写的工作量有多少，但既然别人都写过了为什么不拿、什么都写只会让你成为一个很厉害的写代码的我希望抱大腿的人！</p>
<h2 id="怎么做">怎么做</h2>
<h3 id="jvarkit----java-utilities-for-bioinformatics">jvarkit -- Java
utilities for Bioinformatics</h3>
<p>在BioStar里绝望查找时找到的和当时我的需求最匹配的一个包，呵呵虽然之后还是又改了方向不再使用但还是献上我的感激之情！此包基于java开发，有超多功能
简单总结这个包怎么用：</p>
<ol type="1">
<li><p>先把参考基因组文件拿到！</p></li>
<li><p>jvarkit里的biostar251649
jar包会在原vcf文件的基础上，在vcf的数据部分里的INFO部分加上两个字段：SEQ3_30和SEQ5_30(30是上下游序列的长度，可改)，输出相应的新vcf文件。</p></li>
<li><p>jvarkit里的bioalcidaejdk*
jar包可以把这个新vcf文件里的两个新字段提取出来，和变异信息组成序列</p>
<p>*你以为这是脸滚键盘其实此包是java-based version of awk of
bioinformatics，bio和jdk好理解，Alcidae(海雀科)？因为海雀的英文是auk…
just sooo random! 引用来自文档(女孩你还专门去看了…？)</p></li>
</ol>
<blockquote>
<p><strong>Why this name</strong></p>
<p>As 'bioalcidae' looks like an 'awk' for bioinformatics, we used '<a
target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Alcidae">Alcidae</a>', the taxonomic
Family of the '<a target="_blank" rel="noopener" href="https://en.wikipedia.org/wiki/Auk">auk</a>'
species.</p>
</blockquote>
<p>根据实际调试发现：此包对变异是SNP还是Indel，是双等位还是多等位都不敏感（下面例子可得）。</p>
<p>例子：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><code class="hljs bash"><span class="hljs-comment">## 得到的新vcf</span><br>$ java -jar dist/biostar251649.jar -n 10 -R tests/ref.fa tests/mutations.vcf<br><span class="hljs-comment">##INFO=&lt;ID=SEQ3_10,Number=1,Type=String,Description=&quot;Sequence on the 3&#x27; of mutation&quot;&gt;</span><br><span class="hljs-comment">##INFO=&lt;ID=SEQ5_10,Number=1,Type=String,Description=&quot;Sequence on the 5&#x27; of mutation&quot;&gt;</span><br>(...)<br><span class="hljs-comment">#CHROM  POS ID  REF ALT QUAL    FILTER  INFO    FORMAT  S1  S2  S3  S4</span><br>rotavirus   51  .   A   G   22.55   .   AC1=2;AF1=0.25;BQB=1;DP=944;DP4=849,0,93,0;FQ=23.7972;G3=0.75,0,0.25;HWE=0.033921;MQ=60;MQ0F=0;MQB=1;PV4=1,1,1,1;RPB=0.993129;SEQ3_10=GATGGTAAGC;SEQ5_10=TCTACTCAGC;SGB=-61.9012;VDB=3.53678e-05    GT:PL   0/0:0,255,134   0/0:0,255,127   0/0:0,255,137   1/1:70,255,0<br>rotavirus   91  .   A   T   5.45    .   AC1=1;AF1=0.124963;BQB=0.951201;DP=1359;DP4=1134,0,225,0;FQ=5.8713;MQ=60;MQ0F=0;MQB=1;PV4=1,4.80825e-05,1,1;RPB=0.0393173;SEQ3_10=GTTGTTGCTG;SEQ5_10=TTGAAGCTGC;SGB=-369.163;VDB=0.313337   GT:PL   0/0:0,255,133   0/1:40,0,31 0/0:0,255,134   0/0:0,255,82<br><br><span class="hljs-comment">## 取序列</span><br>java -jar dist/bioalcidaejdk.jar -F VCF -e <span class="hljs-string">&#x27;stream().forEach(V-&gt;println(&quot;&gt;&quot;+V.getContig()+&quot;:&quot;+V.getStart()+&quot;\n&quot;+V.getAttribute(&quot;SEQ5_20&quot;)+&quot;[&quot;+V.getAlleles().stream().map(A-&gt;A.getDisplayString()).collect(Collectors.joining(&quot;/&quot;))+&quot;]&quot;+V.getAttribute(&quot;SEQ3_20&quot;)));&#x27;</span><br><br>&gt;rotavirus:51<br>TGGTCGATTGCTCTATTGAA[A/G]AATTTCCATTGATGGCTAAA <br><span class="hljs-comment"># 这里会把包括参考基因型和变异基因型都列在[]里，所以说对snp/indel，双/多等位不会很敏感</span><br></code></pre></td></tr></table></figure>
<p>详解第二步代码的-e(来自作者在biostars 334253里的回答)</p>
<blockquote>
<p><code>stream()</code> convert to a stream of variant</p>
<p><code>forEach(V-&gt;println("&gt;"</code> for each variant start
writing the fasta header</p>
<p><code>V.getContig()+":"+V.getStart()+"\n"</code> write the fasta
header</p>
<p><code>V.getAttribute("SEQ5_20")</code> write 5' seq</p>
<p><code>["+V.getAlleles().stream().map(A-&gt;A.getDisplayString()).collect(Collectors.joining("/"))+"]"</code>
write all alleles separated with '/'</p>
<p><code>V.getAttribute("SEQ3_20")))</code> write 3' seq</p>
</blockquote>
<p>我自己批量处理vcf的一段代码</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br></pre></td><td class="code"><pre><code class="hljs bash">/#!/bin/bash<br><span class="hljs-comment">#SBATCH --job-name=          # 作业名称</span><br><span class="hljs-comment">#SBATCH --output=        # 标准输出文件</span><br><span class="hljs-comment">#SBATCH --error=         # 标准错误文件</span><br><span class="hljs-comment">#SBATCH --time=                     # 运行时间（格式：HH:MM:SS）</span><br><span class="hljs-comment">#SBATCH --ntasks=1                          # 分配任务数</span><br><span class="hljs-comment">#SBATCH --cpus-per-task=4                   # 每个任务分配的CPU数量（根据需求调整）</span><br><span class="hljs-comment">#SBATCH --mem=16G                           # 分配内存（根据需求调整）</span><br><br><br><span class="hljs-comment"># 设置脚本在遇到错误时停止</span><br><span class="hljs-built_in">set</span> -e<br><br><span class="hljs-comment"># 样品文件路径</span><br>SAMPLE_FILE=<span class="hljs-string">&quot;/path/to/sample/file&quot;</span><br><br><span class="hljs-comment"># VCF 输入文件路径</span><br>VCF_INPUT=<span class="hljs-string">&quot;/path/to/vcf&quot;</span><br><br><span class="hljs-comment"># 基因组参考序列路径</span><br>REFERENCE_GENOME=<span class="hljs-string">&quot;/path/to/fna&quot;</span><br><br><span class="hljs-comment"># JVarkit 工具路径</span><br>JVARKIT_JAR=<span class="hljs-string">&quot;/path/to/jvarkit.jar&quot;</span><br><br><span class="hljs-comment"># 输出目录（可选，建议统一输出到一个目录）</span><br>OUTPUT_DIR=<span class="hljs-string">&quot;/path/to/output/dir&quot;</span><br><span class="hljs-built_in">mkdir</span> -p <span class="hljs-string">&quot;<span class="hljs-variable">$OUTPUT_DIR</span>&quot;</span><br><br><span class="hljs-comment"># 读取每个样品并处理</span><br><span class="hljs-keyword">while</span> IFS= <span class="hljs-built_in">read</span> -r SAMPLE || [[ -n <span class="hljs-string">&quot;<span class="hljs-variable">$SAMPLE</span>&quot;</span> ]]; <span class="hljs-keyword">do</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;开始处理样品: <span class="hljs-variable">$SAMPLE</span>&quot;</span><br><br>    <span class="hljs-comment"># 步骤 1: 使用 bcftools 提取样品 VCF</span><br>    bcftools view -Ov -s <span class="hljs-string">&quot;<span class="hljs-variable">$SAMPLE</span>&quot;</span> <span class="hljs-string">&quot;<span class="hljs-variable">$VCF_INPUT</span>&quot;</span> -o <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>.vcf&quot;</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;生成 <span class="hljs-variable">$&#123;SAMPLE&#125;</span>.vcf&quot;</span><br><br>    <span class="hljs-comment"># 步骤 2: 使用 GATK 选择变异 (主要是想取每个样品的变异)</span><br>    gatk SelectVariants -V <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>.vcf&quot;</span> --exclude-non-variants -O <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>_variant.vcf&quot;</span><br>    <span class="hljs-comment">#echo &quot;生成 $&#123;SAMPLE&#125;_variant.vcf&quot;</span><br><br>    <span class="hljs-comment"># 步骤 3: 使用 JVarkit 提取 flanking regions</span><br>    java -jar <span class="hljs-string">&quot;<span class="hljs-variable">$JVARKIT_JAR</span>&quot;</span> biostar251649 \<br>        -R <span class="hljs-string">&quot;<span class="hljs-variable">$REFERENCE_GENOME</span>&quot;</span> \<br>        <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>.vcf&quot;</span> -n 50 &gt; <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>_flank.vcf&quot;</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;生成 <span class="hljs-variable">$&#123;SAMPLE&#125;</span>_flank.vcf&quot;</span><br><br>    <span class="hljs-comment"># 步骤 4: 使用 JVarkit 生成 FASTA 文件</span><br>    <span class="hljs-comment"># 如果改了上下游长度，这里也记得改</span><br>    java -jar <span class="hljs-string">&quot;<span class="hljs-variable">$JVARKIT_JAR</span>&quot;</span> bioalcidaejdk \<br>        -F VCF -e <span class="hljs-string">&#x27;stream().forEach(V-&gt;println(&quot;&gt;&quot;+V.getContig()+&quot;:&quot;+V.getStart()+&quot;\n&quot;+V.getAttribute(&quot;SEQ5_50&quot;)+&quot;[&quot;+V.getAlleles().stream().map(A-&gt;A.getDisplayString()).collect(Collectors.joining(&quot;/&quot;))+&quot;]&quot;+V.getAttribute(&quot;SEQ3_50&quot;)));&#x27;</span> \<br>        <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>_flank.vcf&quot;</span> &gt; <span class="hljs-string">&quot;<span class="hljs-variable">$&#123;OUTPUT_DIR&#125;</span>/<span class="hljs-variable">$&#123;SAMPLE&#125;</span>_flank.fasta&quot;</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;生成 <span class="hljs-variable">$&#123;SAMPLE&#125;</span>_flank.fasta&quot;</span><br><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;完成处理样品: <span class="hljs-variable">$SAMPLE</span>&quot;</span><br>    <span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;-----------------------------------------&quot;</span><br><br><span class="hljs-keyword">done</span> &lt; <span class="hljs-string">&quot;<span class="hljs-variable">$SAMPLE_FILE</span>&quot;</span><br><br><span class="hljs-built_in">echo</span> <span class="hljs-string">&quot;所有样品处理完成！&quot;</span><br></code></pre></td></tr></table></figure>
<h3 id="vcf2fasta">vcf2fasta</h3>
<p>我猜这个应该是之前用得比较多的一个……？但不知道之前大家用来干什么。和上面的包的主要区别是这个包会根据gff文件里面的基因组特征提取序列，所以这个包的序列提取是以基因组特征为主的，和上面以多态性位点为主不同。</p>
<p>翻找自己的文件发现早就把使用此包的代码删掉了，因为当时用了之后结果……【前情提要：当时的需求是取所有样品某个基因的序列，在有测序数据的前提下其实可以用组装数据来做但我没有！所以就先试着根据有的群体vcf文件和参考序列来取了】不能说是错吧，可能和用的gff文件里面注释有关系，取出来的蛋白质编码基因不是起始密码子开头，影响了后续的计算！回首过往觉得如果要解决的话应该是回头确认gff文件对目标基因的注释到底是从哪开始，对应的具体序列又是什么。虽然软件对CDS的inframe是有参数设定的，但我需要整个基因！当时因为太赶了最后通过最简单也是最难的【问师兄要组装好的序列】解决了呵呵。所以后续如果还需要使用请注意。</p>
<p>怎么用：</p>
<ol type="1">
<li><p>准备参考序列，需要参考基因组的fa文件和gff文件。参考基因组的fa文件需要用samtools进行index；如果你的gff文件非常有个人特色的话，去github看看符不符合软件的要求。vcf文件也要index，但用tabix。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><code class="hljs bash">samtools faidx ref.fa<br>bgzip my_vcf_file.vcf<br>tabix my_vcf_file.vcf.gz<br></code></pre></td></tr></table></figure></li>
<li><p>下载本体、pysam和art包</p></li>
<li><p>最简单的用法，其他参数按-h或者看github吧：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><code class="hljs python">python vcf2fasta.py -f genome.fas -v variants.vcf.gz -g intervals.gff -e CDS<br></code></pre></td></tr></table></figure></li>
<li><p>设定了-e CDS之后会输出所有的CDS，，希望之后能有用</p></li>
</ol>
<h2 id="上面的链接">上面的链接🔗</h2>
<p><a target="_blank" rel="noopener" href="https://github.com/lindenb/jvarkit">jvarkit_github</a></p>
<p><a
target="_blank" rel="noopener" href="https://jvarkit.readthedocs.io/en/latest/">jvarkit_document</a></p>
<p><a
target="_blank" rel="noopener" href="https://github.com/santiagosnchez/vcf2fasta">vcf2fasta</a></p>
<p>BioStars的帖子</p>
<p><a target="_blank" rel="noopener" href="https://www.biostars.org/p/251649/">251649</a></p>
<p><a target="_blank" rel="noopener" href="https://www.biostars.org/p/334253/">334253</a></p>
<h2 id="结束">结束、</h2>
<p>此地的一个好处是写正经笔记的同时女孩还是不由自主散发出一种absent-minded的语言风格，但如果你的名字不是没有人的话你不会在意。</p>

                
              </div>
            
            <hr/>
            <div>
              <div class="post-metas my-3">
  
    <div class="post-meta mr-3 d-flex align-items-center">
      <i class="iconfont icon-category"></i>
      

<span class="category-chains">
  
  
    
      <span class="category-chain">
        
  <a href="/categories/note/" class="category-chain-item">note</a>
  
  

      </span>
    
  
</span>

    </div>
  
  
</div>


              
  

  <div class="license-box my-3">
    <div class="license-title">
      <div>vcf2fasta</div>
      <div>https://patchaven.github.io/2025/01/15/vcf2fasta/</div>
    </div>
    <div class="license-meta">
      
        <div class="license-meta-item">
          <div>作者</div>
          <div>ph-bear</div>
        </div>
      
      
        <div class="license-meta-item license-meta-date">
          <div>发布于</div>
          <div>2025年1月15日</div>
        </div>
      
      
      
        <div class="license-meta-item">
          <div>许可协议</div>
          <div>
            
              
              
                <a class="print-no-link" target="_blank" href="https://creativecommons.org/licenses/by/4.0/">
                  <span class="hint--top hint--rounded" aria-label="BY - 署名">
                    <i class="iconfont icon-cc-by"></i>
                  </span>
                </a>
              
            
          </div>
        </div>
      
    </div>
    <div class="license-icon iconfont"></div>
  </div>



              
                <div class="post-prevnext my-3">
                  <article class="post-prev col-6">
                    
                    
                      <a href="/2025/02/08/robertawhy/" title="roberta模型你在干嘛">
                        <i class="iconfont icon-arrowleft"></i>
                        <span class="hidden-mobile">roberta模型你在干嘛</span>
                        <span class="visible-mobile">上一篇</span>
                      </a>
                    
                  </article>
                  <article class="post-next col-6">
                    
                    
                      <a href="/2024/12/10/flush_slurm/" title="为什么python输出在out文件里不显示啊！！">
                        <span class="hidden-mobile">为什么python输出在out文件里不显示啊！！</span>
                        <span class="visible-mobile">下一篇</span>
                        <i class="iconfont icon-arrowright"></i>
                      </a>
                    
                  </article>
                </div>
              
            </div>

            
          </article>
        </div>
      </div>
    </div>

    <div class="side-col d-none d-lg-block col-lg-2">
      
  <aside class="sidebar" style="margin-left: -1rem">
    <div id="toc">
  <p class="toc-header">
    <i class="iconfont icon-list"></i>
    <span>目录</span>
  </p>
  <div class="toc-body" id="toc-body"></div>
</div>



  </aside>


    </div>
  </div>
</div>





  



  



  



  



  







    

    
      <a id="scroll-top-button" aria-label="TOP" href="#" role="button">
        <i class="iconfont icon-arrowup" aria-hidden="true"></i>
      </a>
    

    
      <div class="modal fade" id="modalSearch" tabindex="-1" role="dialog" aria-labelledby="ModalLabel"
     aria-hidden="true">
  <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
    <div class="modal-content">
      <div class="modal-header text-center">
        <h4 class="modal-title w-100 font-weight-bold">搜索</h4>
        <button type="button" id="local-search-close" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body mx-3">
        <div class="md-form mb-5">
          <input type="text" id="local-search-input" class="form-control validate">
          <label data-error="x" data-success="v" for="local-search-input">关键词</label>
        </div>
        <div class="list-group" id="local-search-result"></div>
      </div>
    </div>
  </div>
</div>

    

    
  </main>

  <footer>
    <div class="footer-inner">
  
  
    <div class="statistics">
  
  

  
    
      <span id="busuanzi_container_site_pv" style="display: none">
        
        <span id="busuanzi_value_site_pv"></span>
         次
      </span>
    
    
      <span id="busuanzi_container_site_uv" style="display: none">
        
        <span id="busuanzi_value_site_uv"></span>
         人
      </span>
    
    

  

</div>

  
  
</div>

  </footer>

  <!-- Scripts -->
  
  <script  src="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.js" ></script>
  <link  rel="stylesheet" href="https://lib.baomitu.com/nprogress/0.2.0/nprogress.min.css" />

  <script>
    NProgress.configure({"showSpinner":false,"trickleSpeed":100})
    NProgress.start()
    window.addEventListener('load', function() {
      NProgress.done();
    })
  </script>


<script  src="https://lib.baomitu.com/jquery/3.6.4/jquery.min.js" ></script>
<script  src="https://lib.baomitu.com/twitter-bootstrap/4.6.1/js/bootstrap.min.js" ></script>
<script  src="/js/events.js" ></script>
<script  src="/js/plugins.js" ></script>





  
    <script  src="/js/img-lazyload.js" ></script>
  




  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/tocbot/4.20.1/tocbot.min.js', function() {
    var toc = jQuery('#toc');
    if (toc.length === 0 || !window.tocbot) { return; }
    var boardCtn = jQuery('#board-ctn');
    var boardTop = boardCtn.offset().top;

    window.tocbot.init(Object.assign({
      tocSelector     : '#toc-body',
      contentSelector : '.markdown-body',
      linkClass       : 'tocbot-link',
      activeLinkClass : 'tocbot-active-link',
      listClass       : 'tocbot-list',
      isCollapsedClass: 'tocbot-is-collapsed',
      collapsibleClass: 'tocbot-is-collapsible',
      scrollSmooth    : true,
      includeTitleTags: true,
      headingsOffset  : -boardTop,
    }, CONFIG.toc));
    if (toc.find('.toc-list-item').length > 0) {
      toc.css('visibility', 'visible');
    }

    Fluid.events.registerRefreshCallback(function() {
      if ('tocbot' in window) {
        tocbot.refresh();
        var toc = jQuery('#toc');
        if (toc.length === 0 || !tocbot) {
          return;
        }
        if (toc.find('.toc-list-item').length > 0) {
          toc.css('visibility', 'visible');
        }
      }
    });
  });
</script>


  <script src=https://lib.baomitu.com/clipboard.js/2.0.11/clipboard.min.js></script>

  <script>Fluid.plugins.codeWidget();</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/anchor-js/5.0.0/anchor.min.js', function() {
    window.anchors.options = {
      placement: CONFIG.anchorjs.placement,
      visible  : CONFIG.anchorjs.visible
    };
    if (CONFIG.anchorjs.icon) {
      window.anchors.options.icon = CONFIG.anchorjs.icon;
    }
    var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
    var res = [];
    for (var item of el) {
      res.push('.markdown-body > ' + item.trim());
    }
    if (CONFIG.anchorjs.placement === 'left') {
      window.anchors.options.class = 'anchorjs-link-left';
    }
    window.anchors.add(res.join(', '));

    Fluid.events.registerRefreshCallback(function() {
      if ('anchors' in window) {
        anchors.removeAll();
        var el = (CONFIG.anchorjs.element || 'h1,h2,h3,h4,h5,h6').split(',');
        var res = [];
        for (var item of el) {
          res.push('.markdown-body > ' + item.trim());
        }
        if (CONFIG.anchorjs.placement === 'left') {
          anchors.options.class = 'anchorjs-link-left';
        }
        anchors.add(res.join(', '));
      }
    });
  });
</script>


  
<script>
  Fluid.utils.createScript('https://lib.baomitu.com/fancybox/3.5.7/jquery.fancybox.min.js', function() {
    Fluid.plugins.fancyBox();
  });
</script>


  <script>Fluid.plugins.imageCaption();</script>

  <script  src="/js/local-search.js" ></script>

  <script defer src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js" ></script>





<!-- 主题的启动项，将它保持在最底部 -->
<!-- the boot of the theme, keep it at the bottom -->
<script  src="/js/boot.js" ></script>


  

  <noscript>
    <div class="noscript-warning">博客在允许 JavaScript 运行的环境下浏览效果更佳</div>
  </noscript>
</body>
</html>
